File: automata_simulator.rb
# Automata & Formal Language Simulator in Ruby

class State
  attr_reader :name
  attr_accessor :accepting
  
  def initialize(name, accepting = false)
    @name = name
    @accepting = accepting
  end
  
  def accepting?
    @accepting
  end
  
  def to_s
    @name
  end
  
  def ==(other)
    @name == other.name
  end
  
  def hash
    @name.hash
  end
  
  alias eql? ==
end

class Transition
  attr_reader :from, :to, :symbol
  
  def initialize(from, to, symbol)
    @from = from
    @to = to
    @symbol = symbol
  end
  
  def matches?(current_state, input_char)
    @from == current_state && @symbol == input_char
  end
  
  def to_s
    "#{@from} --#{@symbol}--> #{@to}"
  end
end

class FiniteStateMachine
  attr_reader :states, :transitions, :initial_state, :current_state, :alphabet
  
  def initialize(states, transitions, initial_state, alphabet)
    @states = states
    @transitions = transitions
    @initial_state = initial_state
    @current_state = initial_state
    @alphabet = alphabet
  end
  
  def reset
    @current_state = @initial_state
  end
  
  def process_string(input)
    reset
    trace = ["Starting at state: #{@current_state}"]
    
    input.each_char do |char|
      unless @alphabet.include?(char)
        trace << "Error: '#{char}' not in alphabet"
        return { accepted: false, trace: trace }
      end
      
      transition = @transitions.find { |t| t.matches?(@current_state, char) }
      
      if transition
        old_state = @current_state
        @current_state = transition.to
        trace << "Read '#{char}': #{old_state} → #{@current_state}"
      else
        trace << "No transition for '#{char}' from #{@current_state}"
        return { accepted: false, trace: trace }
      end
    end
    
    accepted = @current_state.accepting?
    trace << (accepted ? "✓ String ACCEPTED" : "✗ String REJECTED")
    { accepted: accepted, trace: trace }
  end
  
  def add_state(state)
    @states << state unless @states.include?(state)
  end
  
  def add_transition(transition)
    @transitions << transition
  end
  
  def visualize
    puts "\n=== FSM Visualization ==="
    puts "States: #{@states.map(&:name).join(', ')}"
    puts "Accept States: #{@states.select(&:accepting?).map(&:name).join(', ')}"
    puts "Initial State: #{@initial_state}"
    puts "\nTransitions:"
    @transitions.each { |t| puts "  #{t}" }
    puts "========================\n"
  end
end

class NFAConverter
  def self.epsilon_closure(states, nfa_states)
    closure = states.dup
    stack = states.dup
    
    until stack.empty?
      state = stack.pop
      nfa_state = nfa_states[state]
      next unless nfa_state
      
      nfa_state[:epsilon].each do |epsilon_target|
        unless closure.include?(epsilon_target)
          closure << epsilon_target
          stack << epsilon_target
        end
      end
    end
    
    closure
  end
  
  def self.convert_to_dfa(nfa_states, initial_state, accept_states, alphabet)
    dfa_states = []
    dfa_transitions = []
    unmarked_states = [epsilon_closure([initial_state], nfa_states)]
    marked_states = []
    
    until unmarked_states.empty?
      current_dfa_state = unmarked_states.shift
      marked_states << current_dfa_state
      
      is_accepting = !(current_dfa_state & accept_states).empty?
      dfa_state_name = current_dfa_state.sort.join(',')
      dfa_state = State.new(dfa_state_name, is_accepting)
      dfa_states << dfa_state
      
      alphabet.each do |symbol|
        targets = []
        current_dfa_state.each do |nfa_state_name|
          nfa_state = nfa_states[nfa_state_name]
          targets.concat(nfa_state[:transitions][symbol] || []) if nfa_state
        end
        
        next if targets.empty?
        
        target_closure = epsilon_closure(targets, nfa_states)
        unless marked_states.include?(target_closure) || unmarked_states.include?(target_closure)
          unmarked_states << target_closure
        end
        
        target_state_name = target_closure.sort.join(',')
        target_is_accepting = !(target_closure & accept_states).empty?
        target_state = State.new(target_state_name, target_is_accepting)
        
        dfa_transitions << Transition.new(dfa_state, target_state, symbol)
      end
    end
    
    initial_dfa_state = dfa_states.find { |s| s.name == epsilon_closure([initial_state], nfa_states).sort.join(',') }
    FiniteStateMachine.new(dfa_states, dfa_transitions, initial_dfa_state, alphabet)
  end
end

class RegexEngine
  def self.match(pattern, input)
    begin
      regex = Regexp.new("^#{pattern}$")
      !!(input =~ regex)
    rescue RegexpError => e
      puts "Invalid regex pattern: #{e.message}"
      false
    end
  end
  
  def self.match_with_trace(pattern, input)
    matches = match(pattern, input)
    trace = ["Testing pattern: #{pattern}"]
    trace << "Input string: #{input}"
    
    if matches
      trace << "✓ String MATCHES the pattern!"
      trace << "\nStep-by-step analysis:"
      (1..input.length).each do |i|
        partial = input[0...i]
        partial_match = !!(partial =~ Regexp.new("^#{pattern}"))
        trace << "  #{partial}: #{partial_match ? '✓' : '✗'}"
      end
    else
      trace << "✗ String does NOT match the pattern."
    end
    
    { matches: matches, trace: trace }
  end
end

class AutomataSimulator
  def initialize
    @fsm = nil
    @mode = :fsm
  end
  
  def run
    puts "╔═══════════════════════════════════════════════════════════════╗"
    puts "║      Automata & Formal Language Simulator (Ruby)             ║"
    puts "╚═══════════════════════════════════════════════════════════════╝"
    
    create_sample_fsm
    
    loop do
      puts "\n" + "="*60
      puts "Main Menu:"
      puts "1. Test FSM with string"
      puts "2. Test Regular Expression"
      puts "3. Visualize FSM"
      puts "4. Add State"
      puts "5. Add Transition"
      puts "6. Convert NFA to DFA (Demo)"
      puts "7. Exit"
      print "\nSelect option: "
      
      choice = gets.chomp.to_i
      
      case choice
      when 1
        test_fsm
      when 2
        test_regex
      when 3
        @fsm.visualize
      when 4
        add_state
      when 5
        add_transition
      when 6
        demo_nfa_conversion
      when 7
        puts "\nExiting simulator. Goodbye!"
        break
      else
        puts "Invalid option. Please try again."
      end
    end
  end
  
  private
  
  def create_sample_fsm
    q0 = State.new('q0')
    q1 = State.new('q1')
    q2 = State.new('q2', true)
    
    states = [q0, q1, q2]
    transitions = [
      Transition.new(q0, q1, 'a'),
      Transition.new(q1, q2, 'b'),
      Transition.new(q2, q0, 'c')
    ]
    
    @fsm = FiniteStateMachine.new(states, transitions, q0, ['a', 'b', 'c'])
    puts "\n✓ Sample FSM created (accepts strings matching pattern: (abc)*)"
  end
  
  def test_fsm
    print "\nEnter string to test: "
    input = gets.chomp
    
    result = @fsm.process_string(input)
    puts "\n--- Execution Trace ---"
    result[:trace].each { |line| puts line }
  end
  
  def test_regex
    print "\nEnter regex pattern: "
    pattern = gets.chomp
    
    print "Enter test string: "
    input = gets.chomp
    
    result = RegexEngine.match_with_trace(pattern, input)
    puts "\n--- Analysis ---"
    result[:trace].each { |line| puts line }
  end
  
  def add_state
    print "\nEnter state name: "
    name = gets.chomp
    
    print "Is accepting state? (y/n): "
    accepting = gets.chomp.downcase == 'y'
    
    state = State.new(name, accepting)
    @fsm.add_state(state)
    puts "✓ State '#{name}' added successfully."
  end
  
  def add_transition
    print "\nEnter source state: "
    from_name = gets.chomp
    
    print "Enter destination state: "
    to_name = gets.chomp
    
    print "Enter symbol: "
    symbol = gets.chomp
    
    from_state = @fsm.states.find { |s| s.name == from_name }
    to_state = @fsm.states.find { |s| s.name == to_name }
    
    if from_state && to_state
      transition = Transition.new(from_state, to_state, symbol)
      @fsm.add_transition(transition)
      puts "✓ Transition added: #{transition}"
    else
      puts "✗ Error: One or both states not found."
    end
  end
  
  def demo_nfa_conversion
    puts "\n--- NFA to DFA Conversion Demo ---"
    
    nfa_states = {
      'q0' => { epsilon: ['q1'], transitions: { 'a' => ['q0'] } },
      'q1' => { epsilon: [], transitions: { 'b' => ['q2'] } },
      'q2' => { epsilon: [], transitions: {} }
    }
    
    puts "Converting NFA with ε-transitions to DFA..."
    puts "NFA States: q0, q1, q2"
    puts "ε-transitions: q0 → q1"
    puts "Transitions: q0 --a--> q0, q1 --b--> q2"
    
    dfa = NFAConverter.convert_to_dfa(nfa_states, 'q0', ['q2'], ['a', 'b'])
    
    puts "\n✓ Conversion complete!"
    dfa.visualize
  end
end

if __FILE__ == $0
  simulator = AutomataSimulator.new
  simulator.run
end

File: AutomataSimulator.m
#import <Foundation/Foundation.h>

// MARK: - State Interface
@interface FSMState : NSObject
@property (nonatomic, strong) NSString *name;
@property (nonatomic, assign) BOOL isAccepting;

- (instancetype)initWithName:(NSString *)name accepting:(BOOL)accepting;
@end

@implementation FSMState

- (instancetype)initWithName:(NSString *)name accepting:(BOOL)accepting {
    self = [super init];
    if (self) {
        _name = name;
        _isAccepting = accepting;
    }
    return self;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"State(%@%@)", _name, _isAccepting ? @"*" : @""];
}

- (BOOL)isEqual:(id)object {
    if (![object isKindOfClass:[FSMState class]]) return NO;
    FSMState *other = (FSMState *)object;
    return [self.name isEqualToString:other.name];
}

- (NSUInteger)hash {
    return [self.name hash];
}

@end

// MARK: - Transition Interface
@interface FSMTransition : NSObject
@property (nonatomic, strong) FSMState *fromState;
@property (nonatomic, strong) FSMState *toState;
@property (nonatomic, assign) unichar symbol;

- (instancetype)initWithFrom:(FSMState *)from to:(FSMState *)to symbol:(unichar)symbol;
- (BOOL)matchesState:(FSMState *)state withSymbol:(unichar)symbol;
@end

@implementation FSMTransition

- (instancetype)initWithFrom:(FSMState *)from to:(FSMState *)to symbol:(unichar)symbol {
    self = [super init];
    if (self) {
        _fromState = from;
        _toState = to;
        _symbol = symbol;
    }
    return self;
}

- (BOOL)matchesState:(FSMState *)state withSymbol:(unichar)symbol {
    return [self.fromState isEqual:state] && self.symbol == symbol;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ --%C--> %@", 
            self.fromState.name, self.symbol, self.toState.name];
}

@end

// MARK: - Finite State Machine Interface
@interface FiniteStateMachine : NSObject
@property (nonatomic, strong) NSMutableSet<FSMState *> *states;
@property (nonatomic, strong) NSMutableArray<FSMTransition *> *transitions;
@property (nonatomic, strong) FSMState *initialState;
@property (nonatomic, strong) FSMState *currentState;
@property (nonatomic, strong) NSMutableSet *alphabet;

- (instancetype)initWithStates:(NSSet<FSMState *> *)states
                   transitions:(NSArray<FSMTransition *> *)transitions
                  initialState:(FSMState *)initialState
                      alphabet:(NSSet *)alphabet;
- (void)reset;
- (NSDictionary *)processString:(NSString *)input;
- (void)visualize;
@end

@implementation FiniteStateMachine

- (instancetype)initWithStates:(NSSet<FSMState *> *)states
                   transitions:(NSArray<FSMTransition *> *)transitions
                  initialState:(FSMState *)initialState
                      alphabet:(NSSet *)alphabet {
    self = [super init];
    if (self) {
        _states = [states mutableCopy];
        _transitions = [transitions mutableCopy];
        _initialState = initialState;
        _currentState = initialState;
        _alphabet = [alphabet mutableCopy];
    }
    return self;
}

- (void)reset {
    self.currentState = self.initialState;
}

- (NSDictionary *)processString:(NSString *)input {
    [self reset];
    NSMutableArray<NSString *> *trace = [NSMutableArray array];
    [trace addObject:[NSString stringWithFormat:@"Starting at state: %@", self.currentState.name]];
    
    for (NSInteger i = 0; i < input.length; i++) {
        unichar symbol = [input characterAtIndex:i];
        NSString *symbolStr = [NSString stringWithFormat:@"%C", symbol];
        
        if (![self.alphabet containsObject:symbolStr]) {
            [trace addObject:[NSString stringWithFormat:@"Error: '%C' not in alphabet", symbol]];
            return @{@"accepted": @NO, @"trace": trace};
        }
        
        FSMTransition *transition = nil;
        for (FSMTransition *t in self.transitions) {
            if ([t matchesState:self.currentState withSymbol:symbol]) {
                transition = t;
                break;
            }
        }
        
        if (transition) {
            NSString *oldState = self.currentState.name;
            self.currentState = transition.toState;
            [trace addObject:[NSString stringWithFormat:@"Read '%C': %@ → %@", 
                            symbol, oldState, self.currentState.name]];
        } else {
            [trace addObject:[NSString stringWithFormat:@"No transition for '%C' from %@", 
                            symbol, self.currentState.name]];
            return @{@"accepted": @NO, @"trace": trace};
        }
    }
    
    BOOL accepted = self.currentState.isAccepting;
    [trace addObject:accepted ? @"✓ String ACCEPTED" : @"✗ String REJECTED"];
    
    return @{@"accepted": @(accepted), @"trace": trace};
}

- (void)visualize {
    NSLog(@"\n=== FSM Visualization ===");
    NSLog(@"States: %@", [[self.states allObjects] componentsJoinedByString:@", "]);
    
    NSMutableArray *acceptStates = [NSMutableArray array];
    for (FSMState *state in self.states) {
        if (state.isAccepting) {
            [acceptStates addObject:state.name];
        }
    }
    NSLog(@"Accept States: %@", [acceptStates componentsJoinedByString:@", "]);
    NSLog(@"Initial State: %@", self.initialState.name);
    NSLog(@"\nTransitions:");
    for (FSMTransition *t in self.transitions) {
        NSLog(@"  %@", t);
    }
    NSLog(@"========================\n");
}

@end

// MARK: - Regular Expression Engine
@interface RegexEngine : NSObject
+ (BOOL)matchPattern:(NSString *)pattern withInput:(NSString *)input;
+ (NSDictionary *)matchPattern:(NSString *)pattern withInput:(NSString *)input trace:(BOOL)withTrace;
@end

@implementation RegexEngine

+ (BOOL)matchPattern:(NSString *)pattern withInput:(NSString *)input {
    NSError *error = nil;
    NSString *fullPattern = [NSString stringWithFormat:@"^%@$", pattern];
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:fullPattern
                                                                            options:0
                                                                              error:&error];
    if (error) {
        NSLog(@"Regex error: %@", error.localizedDescription);
        return NO;
    }
    
    NSRange range = NSMakeRange(0, input.length);
    NSTextCheckingResult *match = [regex firstMatchInString:input options:0 range:range];
    return match != nil;
}

+ (NSDictionary *)matchPattern:(NSString *)pattern withInput:(NSString *)input trace:(BOOL)withTrace {
    BOOL matches = [self matchPattern:pattern withInput:input];
    NSMutableArray<NSString *> *trace = [NSMutableArray array];
    
    [trace addObject:[NSString stringWithFormat:@"Testing pattern: %@", pattern]];
    [trace addObject:[NSString stringWithFormat:@"Input string: %@", input]];
    
    if (matches) {
        [trace addObject:@"✓ String MATCHES the pattern!"];
        
        if (withTrace) {
            [trace addObject:@"\nStep-by-step analysis:"];
            for (NSInteger i = 1; i <= input.length; i++) {
                NSString *partial = [input substringToIndex:i];
                BOOL partialMatch = [self matchPattern:pattern withInput:partial];
                [trace addObject:[NSString stringWithFormat:@"  %@: %@", partial, partialMatch ? @"✓" : @"✗"]];
            }
        }
    } else {
        [trace addObject:@"✗ String does NOT match the pattern."];
    }
    
    return @{@"matches": @(matches), @"trace": trace};
}

@end

// MARK: - NFA to DFA Converter
@interface NFAConverter : NSObject
+ (FiniteStateMachine *)convertNFAToDFA:(NSDictionary *)nfaStates
                           initialState:(NSString *)initialState
                           acceptStates:(NSSet *)acceptStates
                               alphabet:(NSSet *)alphabet;
@end

@implementation NFAConverter

+ (NSSet *)epsilonClosure:(NSSet *)states withNFAStates:(NSDictionary *)nfaStates {
    NSMutableSet *closure = [states mutableCopy];
    NSMutableArray *stack = [states.allObjects mutableCopy];
    
    while (stack.count > 0) {
        NSString *state = [stack lastObject];
        [stack removeLastObject];
        
        NSDictionary *nfaState = nfaStates[state];
        if (nfaState) {
            NSArray *epsilonTransitions = nfaState[@"epsilon"];
            for (NSString *epsilonTarget in epsilonTransitions) {
                if (![closure containsObject:epsilonTarget]) {
                    [closure addObject:epsilonTarget];
                    [stack addObject:epsilonTarget];
                }
            }
        }
    }
    
    return closure;
}

+ (FiniteStateMachine *)convertNFAToDFA:(NSDictionary *)nfaStates
                           initialState:(NSString *)initialState
                           acceptStates:(NSSet *)acceptStates
                               alphabet:(NSSet *)alphabet {
    NSMutableSet<FSMState *> *dfaStates = [NSMutableSet set];
    NSMutableArray<FSMTransition *> *dfaTransitions = [NSMutableArray array];
    NSMutableArray<NSSet *> *unmarkedStates = [NSMutableArray array];
    NSMutableArray<NSSet *> *markedStates = [NSMutableArray array];
    
    NSSet *initialClosure = [self epsilonClosure:[NSSet setWithObject:initialState] 
                                   withNFAStates:nfaStates];
    [unmarkedStates addObject:initialClosure];
    
    FSMState *initialDFAState = nil;
    
    while (unmarkedStates.count > 0) {
        NSSet *currentDFAStateSet = [unmarkedStates firstObject];
        [unmarkedStates removeObjectAtIndex:0];
        [markedStates addObject:currentDFAStateSet];
        
        BOOL isAccepting = NO;
        for (NSString *stateName in currentDFAStateSet) {
            if ([acceptStates containsObject:stateName]) {
                isAccepting = YES;
                break;
            }
        }
        
        NSString *dfaStateName = [[currentDFAStateSet.allObjects sortedArrayUsingSelector:@selector(compare:)] 
                                 componentsJoinedByString:@","];
        FSMState *dfaState = [[FSMState alloc] initWithName:dfaStateName accepting:isAccepting];
        [dfaStates addObject:dfaState];
        
        if ([currentDFAStateSet isEqualToSet:initialClosure]) {
            initialDFAState = dfaState;
        }
        
        for (NSString *symbolStr in alphabet) {
            unichar symbol = [symbolStr characterAtIndex:0];
            NSMutableSet *targets = [NSMutableSet set];
            
            for (NSString *nfaStateName in currentDFAStateSet) {
                NSDictionary *nfaState = nfaStates[nfaStateName];
                if (nfaState) {
                    NSDictionary *transitions = nfaState[@"transitions"];
                    NSArray *symbolTargets = transitions[symbolStr];
                    if (symbolTargets) {
                        [targets addObjectsFromArray:symbolTargets];
                    }
                }
            }
            
            if (targets.count > 0) {
                NSSet *targetClosure = [self epsilonClosure:targets withNFAStates:nfaStates];
                
                if (![markedStates containsObject:targetClosure] && 
                    ![unmarkedStates containsObject:targetClosure]) {
                    [unmarkedStates addObject:targetClosure];
                }
                
                NSString *targetStateName = [[targetClosure.allObjects sortedArrayUsingSelector:@selector(compare:)] 
                                           componentsJoinedByString:@","];
                BOOL targetIsAccepting = NO;
                for (NSString *stateName in targetClosure) {
                    if ([acceptStates containsObject:stateName]) {
                        targetIsAccepting = YES;
                        break;
                    }
                }
                FSMState *targetState = [[FSMState alloc] initWithName:targetStateName 
                                                             accepting:targetIsAccepting];
                
                FSMTransition *transition = [[FSMTransition alloc] initWithFrom:dfaState 
                                                                            to:targetState 
                                                                        symbol:symbol];
                [dfaTransitions addObject:transition];
            }
        }
    }
    
    return [[FiniteStateMachine alloc] initWithStates:dfaStates 
                                          transitions:dfaTransitions 
                                         initialState:initialDFAState 
                                             alphabet:alphabet];
}

@end

// MARK: - Main Program
int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSLog(@"╔═══════════════════════════════════════════════════════════════╗");
        NSLog(@"║   Automata & Formal Language Simulator (Objective-C)         ║");
        NSLog(@"╚═══════════════════════════════════════════════════════════════╝\n");
        
        // Create sample FSM
        FSMState *q0 = [[FSMState alloc] initWithName:@"q0" accepting:NO];
        FSMState *q1 = [[FSMState alloc] initWithName:@"q1" accepting:NO];
        FSMState *q2 = [[FSMState alloc] initWithName:@"q2" accepting:YES];
        
        NSSet *states = [NSSet setWithObjects:q0, q1, q2, nil];
        NSArray *transitions = @[
            [[FSMTransition alloc] initWithFrom:q0 to:q1 symbol:'a'],
            [[FSMTransition alloc] initWithFrom:q1 to:q2 symbol:'b'],
            [[FSMTransition alloc] initWithFrom:q2 to:q0 symbol:'c']
        ];
        NSSet *alphabet = [NSSet setWithObjects:@"a", @"b", @"c", nil];
        
        FiniteStateMachine *fsm = [[FiniteStateMachine alloc] initWithStates:states
                                                                 transitions:transitions
                                                                initialState:q0
                                                                    alphabet:alphabet];
        
        [fsm visualize];
        
        // Test FSM
        NSLog(@"=== Testing FSM ===\n");
        NSArray *testStrings = @[@"abc", @"ab", @"abcabc", @"xyz"];
        
        for (NSString *testString in testStrings) {
            NSLog(@"Input: \"%@\"", testString);
            NSDictionary *result = [fsm processString:testString];
            for (NSString *line in result[@"trace"]) {
                NSLog(@"%@", line);
            }
            NSLog(@"\n");
        }
        
        // Test Regular Expressions
        NSLog(@"=== Testing Regular Expressions ===\n");
        NSArray *patterns = @[@"(a|b)*c", @"a+b*", @"(ab)+"];
        NSArray *inputs = @[@"aaac", @"abb", @"abab"];
        
        for (NSInteger i = 0; i < patterns.count; i++) {
            NSDictionary *result = [RegexEngine matchPattern:patterns[i] 
                                                   withInput:inputs[i] 
                                                       trace:YES];
            for (NSString *line in result[@"trace"]) {
                NSLog(@"%@", line);
            }
            NSLog(@"\n");
        }
        
        // Demo NFA to DFA conversion
        NSLog(@"=== NFA to DFA Conversion Demo ===\n");
        NSDictionary *nfaStates = @{
            @"q0": @{@"epsilon": @[@"q1"], @"transitions": @{@"a": @[@"q0"]}},
            @"q1": @{@"epsilon": @[], @"transitions": @{@"b": @[@"q2"]}},
            @"q2": @{@"epsilon": @[], @"transitions": @{}}
        };
        
        NSLog(@"Converting NFA with ε-transitions to DFA...");
        FiniteStateMachine *dfa = [NFAConverter convertNFAToDFA:nfaStates
                                                   initialState:@"q0"
                                                   acceptStates:[NSSet setWithObject:@"q2"]
                                                       alphabet:[NSSet setWithObjects:@"a", @"b", nil]];
        
        NSLog(@"\n✓ Conversion complete!");
        [dfa visualize];
    }
    return 0;
}